Add ways to retrieve operations from the ``QkTarget``

raynelfss opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

- Add struct ``QkTargetOperation`` which contains an operation type, a name, the number of qubits, and the parameters of the instruction.
    - The struct can only be returned via the usage of `qk_target_op_get` (name subject to change).
- If a user wanted to retreive the operation instance, in this case, a gate, we can use `qk_target_op_try_gate` to retrieve the instance, it is the user's responsibility to check whether this instance's type is `QkOperationKind_Gate`.


### Details and comments
Fixes #15282. Based on top of ~~#15206~~ and #15208. Will need to be rebased once they merge.

Close #15172

raynelfss committed:
Initial: Add `QkTargetOperation` to the C-API
- Add struct ``QkTargetOperation`` which contains an operation type, a name, the number of qubits, and the parameters of the instruction.
    - The struct can only be returned via the usage of `qk_target_op_get` (name subject to change).
- If a user wanted to retreive the operation instance, in this case, a gate, we can use `qk_target_op_try_gate` to retrieve the instance, it is the user's responsibility to check whether this instance's type is `QkOperationKind_Gate`.

raynelfss committed:
Docs: Add documentation

raynelfss committed:
Test: Add tests

jakelishman commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1253,6 +1254,223 @@ pub struct CInstructionProperties {
     pub error: f64,
 }
 
+/// Representation of an operation identified within the Target.
+#[repr(C)]
+pub struct CTargetOperation {
+    /// The identifier for the current operation.
+    pub op_type: COperationKind,
+    /// The name of the operation.
+    pub name: *mut c_char,
+    /// The number of qubits this operation supports.
+    pub num_qubits: u32,
+    /// The parameters tied to this operation if fixed, as a `double`.
+    pub params: *mut f64,
+    /// The number of parameters supported by this operation.
+    pub num_params: u32,
+}
```
Arguably this should be called `QkTargetOp` and `CTargetOp` to match its function prefix of `qk_target_op`?

raynelfss replied:
Renamed in 087a939c22da59ed4d6146f1778d1d6b2242ae3f

jakelishman commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1253,6 +1254,223 @@ pub struct CInstructionProperties {
+    /// The parameters tied to this operation if fixed, as a `double`.
+    pub params: *mut f64,
+    /// The number of parameters supported by this operation.
+    pub num_params: u32,
```
Our APIs aren't fully functional around exposing `Param` here, but I still think we need to work out a better way of doing things with the (more common) variable parameters thing here.  Perhaps `num_params` is _always_ the number of parameters, but `params` is `null` if the angles aren't fixed?

That precludes supporting a gate with partially fixed parameters, but we also can't specify that in C right now either, so we'd have to revisit anyway.

raynelfss replied:
I will change this to use `QkParam` instead here as it makes more sense.

**Edit:** Maybe I underestimated this a task a little. We can have it be as you described or we can try and expose a `param: *mut *mut Param` to C, but it would look a little dirty. It may be easier to just do as you said due to how little time we have.

raynelfss replied:
From offline discussion:

> Would you rather have `QkTargetOp` expose parameters as:
> - `param: *mut f64` and `num_param: u32` but `param` is `NULL` if no fixed parameters are specified while `num_params` always shows whatever number of parameters exist.
> - Another variant of the one above but with `num_param` representing only the fixed parameters.
> - `param: *mut *mut Param` and `num_param: u32`. In this case I'm confused as to what to do:
>     - We could take advantage of `smallvec` and turn it directly into a pointer. However, we don't expose `QkParam` as a struct but rather a pointer here. Could we perhaps just give the pointer and cast it somehow?
>     - We could make an array with the parameters as pointer and then turn that into a pointer too. (very messy).

Cryoris replied:
It feels strange to invent a random placeholder `QkParam` if the user never instantiated it, so I would probably lean towards option 1 you describe here: `num_params` describes the number of parameters and if `params` is not NULL then it's readable for `num_params` elements.

jakelishman commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1253,6 +1254,223 @@ pub struct CInstructionProperties {
     pub error: f64,
 }
 
+/// Representation of an operation identified within the Target.
+#[repr(C)]
+pub struct CTargetOperation {
+    /// The identifier for the current operation.
+    pub op_type: COperationKind,
+    /// The name of the operation.
+    pub name: *mut c_char,
+    /// The number of qubits this operation supports.
+    pub num_qubits: u32,
+    /// The parameters tied to this operation if fixed, as a `double`.
+    pub params: *mut f64,
+    /// The number of parameters supported by this operation.
+    pub num_params: u32,
+}
```
We should document that it's undefined behaviour to write to or manually free any of the fields here, for a consumer.

raynelfss replied:
Added documentation in 087a939c22da59ed4d6146f1778d1d6b2242ae3f

jakelishman commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1253,6 +1254,223 @@ pub struct CInstructionProperties {
+    target: *const Target,
+    index: usize,
+    op_kind: *mut CTargetOperation,
+) -> ExitCode {
```
I'm fine with `op_get` here.  `op_info`'s another choice if you prefer to avoid `_get`.

If I'm reading this correctly, the only error return is the index, which I think we've agreed in principle to just treat as a panic in the C API, at least until we're in a position to add some sort of "checked" API in addition.  We can probably just remove the `ExitCode` return here, I think?

raynelfss replied:
Removed `Exitcode` in 087a939c22da59ed4d6146f1778d1d6b2242ae3f

jakelishman commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1253,6 +1254,223 @@ pub struct CInstructionProperties {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let target_borrowed = unsafe { const_ptr_as_ref(target) };
+
+    if let Some(operation) = target_borrowed.get_op_by_index(index) {
```
If we get rid of `ExitCode` this can be `expect`, but if we keep it, there's the possiblity of
```rust
let Some(oepration) = target_borrowed.get_op_by_index(index) else {
    return ExitCode::IndexError;
};
```
to reduce nesting.

jakelishman commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1253,6 +1254,223 @@ pub struct CInstructionProperties {
+pub unsafe extern "C" fn qk_target_op_try_gate(
+    target: *const Target,
+    index: usize,
+) -> StandardGate {
```
The name `_try` is probably not so appropriate, since we panic if it's not a gate!  Just `qk_target_op_gate` is probably ok?

raynelfss replied:
Renamed into `qk_target_op_gate` in 087a939c22da59ed4d6146f1778d1d6b2242ae3f

jakelishman commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1253,6 +1254,223 @@ pub struct CInstructionProperties {
     pub error: f64,
 }
 
+/// Representation of an operation identified within the Target.
+#[repr(C)]
+pub struct CTargetOperation {
+    /// The identifier for the current operation.
+    pub op_type: COperationKind,
+    /// The name of the operation.
+    pub name: *mut c_char,
+    /// The number of qubits this operation supports.
+    pub num_qubits: u32,
+    /// The parameters tied to this operation if fixed, as a `double`.
+    pub params: *mut f64,
+    /// The number of parameters supported by this operation.
+    pub num_params: u32,
+}
```
What's our plan for representing variadics in this?  Perhaps setting `num_qubits` to `u32::MAX` and a helper method `qk_target_op_is_variadic`?

raynelfss replied:
Should I add a `Variadic` kind for `COperationKind`?

jakelishman replied:
At the moment, I think it's just control-flow operations that, in Python space, we put in as just the class (not an instance) and they're then variadic over any parameters and qubits.   Technically the Python-space model allows for _any_ operation to be like that, and if the C API is going to allow exposure of `Target`s created in Python space, we need to have a plan for how we'll handle those, even if it's not fully implemented immediately.

jakelishman reviewed (commented):
Aside from the API questions about how we handle symbolic vs fixed parameters and variadics, this generally looks like a solid step to me, thanks Ray.

jakelishman commented:
Also I think we need a release note?

raynelfss committed:
Fix: initial review comments
- Rename `QkTargetOperation` to `QkTargetOp`.
- Document that users should not tamper with `CTargetOp` mutably.
- Remove `ExitCode` return from `qk_target_op_get`.
- Rename `target_op_try_gate` to `target_op_gate`.

raynelfss committed:
Docs: Add release note.

raynelfss committed:
Fix: Use actual number of parameters for `QkTargetOp`,

raynelfss committed:
Add: Support for returning variadics
- To support variadics, we added support to retrieve operations typed `QkOperationKind::PythonOp` and setting the number of qubits and parameters as `u32::MAX`.

Cryoris commented on crates/cext/src/transpiler/target.rs:
````diff
@@ -1326,6 +1327,239 @@ pub struct CInstructionProperties {
+///
+/// # Example
+/// ```c
+///     QkTarget *target = qk_target_new(5);
````
Tiny and optional nit: the indents here are not needed

raynelfss replied:
Re-indented in ecc86a31ae142c315614b5551c6233918648ee4a

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1326,6 +1327,239 @@ pub struct CInstructionProperties {
+/// store ``QkTargetOp`` in.
+#[unsafe(no_mangle)]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_op_get(
```
I'd have a mild preference for calling this `qk_target_get_op` in alignment with `qk_circuit_get_instruction` but not a required change if you think this makes more sense as is

raynelfss replied:
The reason why we didn't call it that is because we have a couple operations prefixed as `qk_target_op_*` so we thought it might make sense to continue on with that format.

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1326,6 +1327,239 @@ pub struct CInstructionProperties {
+
+/// @ingroup QkTarget
+/// Retrieves information about an operation in the Target via index.
+/// If the index is not present, this function will panic.
```
Would be nice to link to `qk_target_num_instructions` which can be used to check what the maximum allowed index is

raynelfss replied:
Added a callback ecc86a31ae142c315614b5551c6233918648ee4a

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1326,6 +1327,239 @@ pub struct CInstructionProperties {
+                    })
+                    .collect(),
+            );
+            // Safety: As per documentation, `out_op` is a pointer to a sufficient allocation.
```
Another small nit: all inline safety comments start in screaming case (ok I found one occurrence which doesn't but let that be the only outlier lol)

raynelfss replied:
Your nit is my command ecc86a31ae142c315614b5551c6233918648ee4a

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1326,6 +1327,239 @@ pub struct CInstructionProperties {
+                    StandardInstruction::Reset => COperationKind::Reset,
+                },
+                qiskit_circuit::operations::OperationRef::Unitary(_) => COperationKind::Unitary,
+                _ => panic!(
```
I probably missed this but why are `PauliProductMeasurement` and `ControlFlow` not supported here?

raynelfss replied:
Not yet, so we don't have to add them.

**Edit:** We currently have no way of adding those to the `Target` we currently only support Standard Gates, Reset and Measure. We have not added support for anything else so far. But when we do we will modify this.

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1326,6 +1327,239 @@ pub struct CInstructionProperties {
+/// @param target A pointer to the Target instance.
+/// @param index The index at which the operation is located.
+///
+/// @return the QkGate instance in said index.
```
```suggestion
/// @param target A pointer to the ``Target`` instance.
/// @param index The index at which the operation is located.
///
/// @return The ``QkGate`` enum in said index.
```

raynelfss replied:
Added in ecc86a31ae142c315614b5551c6233918648ee4a

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1326,6 +1327,239 @@ pub struct CInstructionProperties {
+///     }
+///
+///     // Clean up after you're done.
+///     qk_target_op_clear(op);
```
```suggestion
///     qk_target_op_clear(&op);
```

raynelfss replied:
Added in ecc86a31ae142c315614b5551c6233918648ee4a

raynelfss committed:
Fix: Review items

Co-authored-by: Julien Gacon <gaconju@gmail.com>

raynelfss committed:
Chore: Rename `COperationKind::PythonOp` to `Unknwon`.

jakelishman commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1326,6 +1327,240 @@ pub struct CInstructionProperties {
+            .expect("The string should be UTF-8 encoded.")
+            .into_raw();
+            let num_params = operation.operation.num_params();
+            let mut params: Option<Box<[f64]>> = (num_params > 0).then_some(
+                operation
+                    .params_view()
+                    .iter()
+                    .filter_map(|param| match param {
+                        Param::Float(number) => Some(*number),
+                        _ => None,
+                    })
+                    .collect(),
```
The logic here doesn't result in a boxed slice that's _necessarily_ `num_params` entries long; it's only valid in C because we didn't provide a way to only specify partially fixed parameters yet.  Once `Target`s can be created (say) by the Python interface and given to C (e.g. #15337), we can't get away with that kind of assumption in the C API any more, so we'll have to handle it.

I'm ok to merge this as-is (and I don't want to go through another CI cycle), but let's take a follow-up issue to (for example) use `f64::NAN` as a marker for "arbitrary" so the slice is definitely going to be the correct length and C space will be able to tell.

jakelishman reviewed (approved):
Thanks for this Ray and Julien.  I think the API's in a place here where we can move to merge - I left a comment for a follow-up that we can fixed without it being a breaking change (because anything that triggered behaviour that'll be different would necessarily cause bugs in the Rust/C code handling it).
